crates/astria-sequencer-relayer/src/relayer/write/conversion.rs

```diff
@@ -1,142 +1,610 @@
+use std::{
+    collections::{
+        BTreeSet,
+        HashMap,
+        HashSet,
+    },
+    pin::Pin,
+    task::Poll,
+};
+
 use astria_core::{
     brotli::compress_bytes,
+    generated::sequencerblock::v1alpha1::{
+        CelestiaHeader,
+        CelestiaHeaderList,
+        CelestiaRollupData,
+        CelestiaRollupDataList,
+    },
     primitive::v1::RollupId,
 };
-use astria_eyre::eyre::{
-    self,
-    WrapErr as _,
-};
 use celestia_types::{
     nmt::Namespace,
     Blob,
 };
-use prost::Message as _;
+use futures::Future;
+use pin_project_lite::pin_project;
 use sequencer_client::SequencerBlock;
 use tendermint::block::Height as SequencerHeight;
-use tracing::debug;
-
-use crate::{
-    metrics_init,
-    IncludeRollup,
+use tracing::{
+    trace,
+    warn,
 };
 
-// allow: the signature is dictated by the `serde(serialize_with = ...)` attribute.
-#[allow(clippy::trivially_copy_pass_by_ref)]
-fn serialize_height<S>(height: &SequencerHeight, serializer: S) -> Result<S::Ok, S::Error>
+use crate::IncludeRollup;
+
+/// The maximum permitted payload size in bytes that relayer will send to Celestia.
+///
+/// Taken as half the maximum block size that Celestia currently allows (2 MB at the moment).
+const MAX_PAYLOAD_SIZE_BYTES: usize = 1_000_000;
+
+/// The maximum size that a payload can have with a single sequencer block included.
+const MAX_EXCEPTIONALLY_LARGE_PAYLOAD_BYTES: usize = 1_500_000;
```

Comment:
We could maybe have a const for max Celestia size and derive these two consts from that one?  Should make it slighter harder to get wrong if/when the max size specified by Celestia changes.

Reply:
Removed this as part of d913803